File: Problem/By section/Two pointer/Problem 345/problem_345.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    unordered_set<char> vowels{'a', 'e', 'i', 'o', 'u',
                                'A', 'E', 'I', 'O', 'U'};

    string reverseVowels(string s) {
        int left = 0;
        int right = s.length() - 1;
        string result(s);

        while(left < right) {
            if(vowels.find(s[left]) == vowels.end()) {
                left ++;
            }else if(vowels.find(s[right]) == vowels.end()){
                right --;
            }else {
                char temp = s[left];
                result[left++] = s[right];
                result[right--] = temp;
            }
        }

        return result;
    }
};

int main(int argc, char** argv) {
    Solution solution;
    std::cout << solution.reverseVowels("leetcode") << std::endl;
    return 0;
}

File: Problem/By section/Two pointer/Problem 345/problem_345.java
import java.util.*;
class Solution{
    // use HashSet to achieve O(1) for determining whether a letter is a vowel
    private final static HashSet<Character> Vowels = new HashSet<>(
        Arrays.asList('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'));
    
    // time: O(s.length), space: O(1)
    public static String reverseVowels(String s) {
        if(s.length() == 0) return "";
        if(s.length() == 1) return s;
        
        //two pointers
        int i = 0;
        int j = s.length() - 1;
        char[] ans = s.toCharArray();
        while(i < j){
            char temp_1 = ans[i];
            char temp_2 = ans[j];
            
            // find the left i that is a voewl
            if(!Vowels.contains(temp_1)){
                i ++;
            }else if(!Vowels.contains(temp_2)){
                // find right j that is a voewl
                j --;
            }else{
                // swap them, and continue
                ans[i ++] = temp_2;
                ans[j --] = temp_1;
            }
        }
        
        return new String(ans);
    }

}